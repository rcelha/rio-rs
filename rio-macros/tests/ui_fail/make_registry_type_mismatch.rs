// TODO create ui_fail for this
//
use std::sync::Arc;

use async_trait::async_trait;
use rio_rs::prelude::*;
use serde::{Deserialize, Serialize};

use rio_macros::make_registry;

#[derive(Default, Debug, WithId, TypeName)]
struct TestService {
    id: String,
}

impl ServiceObjectStateLoad for TestService {}
impl ServiceObject for TestService {}

#[derive(TypeName, Message, Debug, Deserialize, Serialize)]
pub struct Ping {
    pub ping_id: String,
}

#[derive(TypeName, Message, Debug, Deserialize, Serialize)]
pub struct Pong {
    pub ping_id: String,
}

#[async_trait]
impl Handler<Ping> for TestService {
    type Returns = Pong;
    type Error = NoopError;

    async fn handle(
        &mut self,
        message: Ping,
        _app_data: Arc<AppData>,
    ) -> Result<Self::Returns, Self::Error> {
        Ok(Pong {
            ping_id: message.ping_id,
        })
    }
}

make_registry! {
    TestService: [
        Ping => (Ping, NoopError),
    ]
}

fn main() {
    let _registry = server::registry();
}
