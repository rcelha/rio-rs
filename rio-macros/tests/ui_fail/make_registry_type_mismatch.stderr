error[E0271]: type mismatch resolving `<TestService as Handler<Ping>>::Returns == Ping`
  --> tests/ui_fail/make_registry_type_mismatch.rs:45:1
   |
45 | / make_registry! {
46 | |     TestService: [
47 | |         Ping => (Ping, NoopError),
48 | |     ]
49 | | }
   | |_^ expected `Ping`, found `Pong`
   |
note: required by a bound in `assert_handler_type`
  --> tests/ui_fail/make_registry_type_mismatch.rs:45:1
   |
45 | / make_registry! {
46 | |     TestService: [
47 | |         Ping => (Ping, NoopError),
48 | |     ]
49 | | }
   | |_^ required by this bound in `assert_handler_type`
   = note: this error originates in the macro `make_registry` (in Nightly builds, run with -Z macro-backtrace for more info)
